src/Database/Connection.php

```diff
@@ -135,11 +135,14 @@ public function __construct(array $config)
     {
         $this->_config = $config;
 
-        $driver = '';
-        if (!empty($config['driver'])) {
-            $driver = $config['driver'];
-        }
-        $this->setDriver($driver, $config);
+        $driverConfig = array_diff_key($config, array_flip([
+            'name',
+            'driver',
+            'log',
+            'cacheMetaData',
+            'cacheKeyPrefix',
+        ]));
+        $this->_driver = $this->createDriver($config['driver'] ?? '', $driverConfig);
```

Comment:
The datasource config is a combined config for connection and driver which makes this a little awkward.